File: 카카오 겨울인턴2019/불량 사용자.js

function solution(user_id, banned_id) {
    // 해당 유저 만들기 { 'fr*d*': [ 'frodo', 'fradi' ], 'abc1**': [ 'abc123' ] }
    let banDict = {}
    for (const ban of banned_id){
        console.log(ban)
        const regex = new RegExp(ban.replace(/\*/g, '.'));
        const matches = user_id.filter(user =>  // [ 'frodo', 'fradi' ]  [ 'abc123' ]
        regex.exec(user) && regex.exec(user)[0].length === user.length
        );
        banDict[ban] = matches
    }

    // 조합 구하기
    let arr = banned_id.map(item => banDict[item]) // [ [ 'frodo', 'fradi' ], [ 'abc123' ] ]
    let comb = allPossibleCases(arr) // [ [ 'frodo', 'abc123' ], [ 'fradi', 'abc123' ] ]
    
    // 중복 제거
    let answer = []
    let length = banned_id.length
    for (const item of comb){
        let temp = new Set(item) // 중복제거
        if ([...temp].length === length){
            let removed = [...temp].sort().join('') // 스트링화해서 중복제거
            if (!answer.includes(removed)) answer.push(removed)
        }
    }
    return answer.length
}

function allPossibleCases(arr) {
        if (arr.length == 1) {
            return arr[0];
        } 
        else {
            var result = [];
            var allCasesOfRest = allPossibleCases(arr.slice(1));  // recur with the rest of array
            for (var i = 0; i < allCasesOfRest.length; i++) {
                for (var j = 0; j < arr[0].length; j++) {
                    let comb = (arr[0][j] + ',' + allCasesOfRest[i]);
                    result.push(comb.split(','))
                }
            }
        //console.log(result)
        return result;
        }
}

File: 카카오 겨울인턴2019/불량 사용자.py
from collections import defaultdict
import itertools

def solution(user_id, banned_id):
    cand = defaultdict(set) # {'fr*d*': {'frodo', 'fradi'}, 'abc1**': {'abc123'}}
    count = 0
    
    for bid in banned_id:
        for uid in user_id:
            if len(uid) != len(bid): # 길이가 다르면 스킵
                continue
            for u, b in zip(uid, bid): # zip으로 비교
                if u != b and b != '*':
                    break;
            else: # 다 만족했을 때
                cand[bid].add(uid)

    temp = [] # [{'frodo', 'fradi'}, {'abc123'}]
    for bid in banned_id: 
        if cand[bid]: # 해당 아이디가 있다면
            temp.append(cand[bid])
            
    temp2 = itertools.product(*temp) # [('fradi', 'abc123'), ('frodo', 'abc123')]
    result = []
    for x in map(set, list(temp2)): # 중복제거
        if len(x) == len(temp) and set(x) not in result: # 중복된게 없고 result에 없으면
            result.append(set(x))
            count += 1

    return count
